File: app/javascript/channels/private_chat_channel.js
import consumer from "./consumer";

let chatChannel;

// Expose openChat as a global function
window.openChat = function(recipientId, recipientUsername) {
  document.getElementById('chat-modal').style.display = 'block';
  document.getElementById('chat-recipient-username').innerText = recipientUsername;

  // Connect to the chat channel
  chatChannel = consumer.subscriptions.create(

    { channel: "PrivateChatChannel", user_id: USER_ID, recipient_id: recipientId },
    {
      connected() {
        // Called when the subscription is ready for use
      },

      disconnected() {
        // Called when the subscription has been terminated by the server
      },

      received(data) {
        // Append received message to chat
        const messages = document.getElementById('chat-messages');
        messages.innerHTML += `<p><strong>${data.user}:</strong> ${data.message}</p>`;
        messages.scrollTop = messages.scrollHeight; // Scroll to the bottom
      },

      send_message(message) {
        const messages = document.getElementById('chat-messages');
        messages.innerHTML += `<p><strong>You:</strong> ${message}</p>`;
        messages.scrollTop = messages.scrollHeight; // Scroll to the bottom
        this.perform('send_message', { message: message, recipient_id: recipientId, user_id: USER_ID });
      }
    }
  );
}

// Attach event listener for sending chat messages
document.addEventListener('DOMContentLoaded', () => {
  document.getElementById('send-chat').addEventListener('click', () => {
    const messageInput = document.getElementById('chat-input');
    const message = messageInput.value;

    if (chatChannel) {
      chatChannel.send_message(message);
      messageInput.value = ''; // Clear input
    }
  });
});

// Function to close chat modal
window.closeChat = function() {
  document.getElementById('chat-modal').style.display = 'none';
  if (chatChannel) {
    chatChannel.unsubscribe();
  }
}

File: app/channels/private_chat_channel.rb
class PrivateChatChannel < ApplicationCable::Channel
  def subscribed
    # Stream from a specific chat room based on user IDs
    stream_from "private_chat_#{params[:user_id]}_#{params[:recipient_id]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def send_message(data)

    Rails.logger.fatal "^^^^^^^^^^^^^^^^^data^^^^^^^^^^^^^^^^^^^^^^#{data}"
    # Broadcast the message to the specific chat room
    ActionCable.server.broadcast("private_chat_#{data['recipient_id']}_#{data['user_id']}", { message: data['message'], user: data['user_id'] })
  end
end
